### DISCUSSION TITLE
Fixed toolbar in RichText Lexical editor.

### DISCUSSION THREAD

**Discussion by revnelson** [payloadcms/payload#3977] [Q&A] [answered]
I love the lexical editor's floating toolbar for most cases, but is there a way to have the toolbar fixed to the top of the RichText box? There are a few places where it would make more sense , such as fields where inserting an icon in the middle of text is more common than modifying something that has been selected. Currently it seems like the only menus are the slash menu for a new section, or when selecting existing text.

**Comment by johngrantdev** [comment]
While I'm quite happy with using the more modern context menu for my projects, I thought I'd dump some extra arguments for the traditional toolbar here:
- Familiarity for end users - the context menu and + menu hide features by default. This provides a clean and efficient way of working for frequent/power users but the slight learning curve may be challenging or annoying for infrequent users.
- When transitioning from slate to lexical, you may not want to initially deal with a major change to the UI.

**Comment by patricsgamma** [comment]
One additional weakness of the floating toolbar we experience is.

### Our Feature
Some of our Text is sent to a TextToSpeech service to be voiced out, we can easily do this with hooks.
Now we also have a [Substitution alias](https://cloud.google.com/text-to-speech/docs/ssml#sub) feature implemented that lets us replace certain words of a text to be spoken differently.
A next feature that would be nice is to have the possibility to add Audio files inline to be added and sent along to the TTS service.

### Missing Payload lexical feature
For a use-case like this, we would need to add a marker at an arbitrary text position. But we have no toolbar to achieve this.
Right now we have to create a random character. This so we can select this character which opens the toolbar, then we can add a our desired marker node instead of this char. 
But I can't see how to get a toolbar to open at a desired curser position, without such a hack.

**Comment by pvaladez** [comment]
I would also like to see a traditional fixed toolbar, like the official Lexical playground has: https://playground.lexical.dev/

Additionally I would like there to be a border and/or different color background on the rich text editing area.  Right now the Lexical editor in Payload just makes the background the same color as everything around it, and it's hard to tell where the rich text editing area starts and ends.  I believe this in combination with the fact that there is no fixed toolbar will cause a lot of confusion and questions for end users.

**Comment by AlessioGr** [comment, marked as answer]
Thank you for your feedback! We listen to our community, so we've just added a fixed toolbar to 3.0-beta.25! :tada:  You can test it out by enabling the `FixedToolbarFeature` like this

```ts
editor: lexicalEditor({
  features: ({ defaultFeatures }) => [
    ...defaultFeatures,
    FixedToolbarFeature(),
  ],
})
```
![CleanShot 2024-05-06 at 13 43 04@2x](https://github.com/payloadcms/payload/assets/70709113/9b72d1b5-4883-4a67-8301-fe605f9edf32)

It's fully-configurable and you can add your own buttons, dropdowns or any other react component to it - just like with our floating toolbar!

While it is an opt-in feature, so by default, only the inline toolbar is enabled, it will receive the same level of support as the default inline toolbar.

Additionally, the default inline toolbar has been extracted into a `InlineToolbarFeature`. This means that, if you want to, you can now disable the inline toolbar.

**Reply by SimonVreman** [reply]
@AlessioGr not sure if known, but it has some overflow issues on smaller screens, when using live preview for example.
<img width="1410" alt="image" src="https://github.com/payloadcms/payload/assets/35310171/f76cf32a-6db5-468b-a4df-fea057955600">

**Reply by jalmanzaa** [reply]
I have a doubt, ¿have you already implemented an option to change the color of a specific text?, that is to say; select a text and have it have x color, something similar to the link below.

**Comment by neeraj-9** [comment]
Anyone facing the issue with `FixedToolbarFeature`? After enabling this feature I still can't see the fixed tool bar. My payload version is 3.56.0

**Reply by GermanJablo** [reply]
Many tests in our CI rely on fixedToolbar, so it's probably a configuration issue. Feel free to open a Issue or share a repro

**Reply by Sowmo0509** [reply]
Still having the issue. Any fixes?

**Reply by Sowmo0509** [reply]
> Still having the issue. Any fixes?

Okay I think I solved it lol. I was using a different instance, I was not seeing this, but when removing the editor and keeping only `{
            name: 'content',
            type: 'richText',
            required: true,
        },`
This automatically worked.
